grpc-circuitbreaker-utils/src/main/java/org/hypertrace/circuitbreaker/grpcutils/resilience/ResilienceCircuitBreakerInterceptor.java

```diff
@@ -0,0 +1,147 @@
+package org.hypertrace.circuitbreaker.grpcutils.resilience;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.github.resilience4j.circuitbreaker.CircuitBreaker;
+import io.github.resilience4j.circuitbreaker.CircuitBreakerConfig;
+import io.github.resilience4j.circuitbreaker.CircuitBreakerRegistry;
+import io.grpc.CallOptions;
+import io.grpc.Channel;
+import io.grpc.ClientCall;
+import io.grpc.ForwardingClientCall;
+import io.grpc.ForwardingClientCallListener;
+import io.grpc.Metadata;
+import io.grpc.MethodDescriptor;
+import io.grpc.Status;
+import java.time.Clock;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import lombok.extern.slf4j.Slf4j;
+import org.hypertrace.circuitbreaker.grpcutils.CircuitBreakerConfiguration;
+import org.hypertrace.circuitbreaker.grpcutils.CircuitBreakerInterceptor;
+import org.hypertrace.core.grpcutils.context.RequestContext;
+
+@Slf4j
+public class ResilienceCircuitBreakerInterceptor extends CircuitBreakerInterceptor {
+
+  private final CircuitBreakerRegistry resilicenceCircuitBreakerRegistry;
+  private final Map<String, CircuitBreakerConfig> resilienceCircuitBreakerConfigMap;
+  private final ResilienceCircuitBreakerProvider resilienceCircuitBreakerProvider;
+  private final CircuitBreakerConfiguration<?> circuitBreakerConfiguration;
+  private final Clock clock;
+
+  public ResilienceCircuitBreakerInterceptor(
+      CircuitBreakerConfiguration<?> circuitBreakerConfiguration, Clock clock) {
+    this.circuitBreakerConfiguration = circuitBreakerConfiguration;
+    this.clock = clock;
+    this.resilienceCircuitBreakerConfigMap =
+        ResilienceCircuitBreakerConfigParser.getCircuitBreakerConfigs(
+            circuitBreakerConfiguration.getCircuitBreakerThresholdsMap());
+    this.resilicenceCircuitBreakerRegistry =
+        new ResilienceCircuitBreakerRegistryProvider(resilienceCircuitBreakerConfigMap)
+            .getCircuitBreakerRegistry();
+    this.resilienceCircuitBreakerProvider =
+        new ResilienceCircuitBreakerProvider(
+            resilicenceCircuitBreakerRegistry, resilienceCircuitBreakerConfigMap);
+  }
+
+  @VisibleForTesting
+  public ResilienceCircuitBreakerInterceptor(
```

Comment:
Visible for testing is an antipattern for testing old code - do we really need to test this with a separate registry? It's either an internal impl detail and we don't need to test it separately, or it should be passed in for the regular construction path.

Reply:
We need to mock these classes for testing. One alternative way is creating factory and client code will look like 
`new ResilienceCircuitBreakerInterceptor(
    config,
    clock,
    CircuitBreakerDependencyFactory.createRegistry(config.getDefaultThresholds()),
    CircuitBreakerDependencyFactory.createProvider(
        registry, 
        ResilienceCircuitBreakerConfigConverter.getCircuitBreakerConfigs(config.getCircuitBreakerThresholdsMap())
    )
);
`

Reply:
Agree, to mitigate the behaviour of different scenarios, we need to mock registry and provider. But is it ok in the client side we need to pass `CircuitBreakerDependencyFactory.createRegistry(config.getDefaultThresholds()), CircuitBreakerDependencyFactory.createProvider( registry,  ResilienceCircuitBreakerConfigConverter.getCircuitBreakerConfigs(config.getCircuitBreakerThresholdsMap())  as constructor arguments` these as well. Won't these unnecessary to consumer ?

Reply:
Added a factory and fixed it.